gctscript/vm/vm.go

```diff
@@ -0,0 +1,202 @@
+package vm
+
+import (
+	"context"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/d5/tengo/script"
+	"github.com/gofrs/uuid"
+	"github.com/thrasher-corp/gocryptotrader/database/repository/audit"
+	"github.com/thrasher-corp/gocryptotrader/gctscript/modules/loader"
+	log "github.com/thrasher-corp/gocryptotrader/logger"
+)
+
+func newVM() *VM {
+	newUUID, err := uuid.NewV4()
+	if err != nil {
+		log.Error(log.GCTScriptMgr, Error{
+			Action: "New -> UUID",
+			Cause:  err,
+		})
+		return nil
+	}
+
+	if GCTScriptConfig.Verbose {
+		log.Debugln(log.GCTScriptMgr, "New GCTScript VM created")
+	}
+
+	audit.Event(newUUID.String(), AuditEventName, "virtual machine created")
+
+	return &VM{
+		ID:     newUUID,
+		Script: pool.Get().(*script.Script),
+	}
+}
+
+// Load parses and creates a new instance of tengo script vm
+func (vm *VM) Load(file string) error {
+	if !GCTScriptConfig.Enabled {
+		return &Error{
+			Action: "Load",
+			Cause:  ErrScriptingDisabled,
+		}
+	}
+
+	if GCTScriptConfig.Verbose {
+		log.Debugf(log.GCTScriptMgr, "Loading script: %v", file)
+	}
+
+	f, err := os.Open(file)
+	if err != nil {
+		return &Error{
+			Action: "Load -> Open",
+			Script: file,
+			Cause:  err,
+		}
+	}
+
+	code, err := ioutil.ReadAll(f)
+	if err != nil {
+		return &Error{
+			Action: "Load -> Read",
+			Script: file,
+			Cause:  err,
+		}
+	}
+
+	vm.file = f.Name()
+	vm.Name = vm.shortName(file)
+	vm.Script = script.New(code)
+	vm.Script.SetImports(loader.GetModuleMap())
+
+	if GCTScriptConfig.AllowImports {
+		if GCTScriptConfig.Verbose {
+			log.Debugf(log.GCTScriptMgr, "file imports enabled for vm: %v", vm.ID)
+		}
+		vm.Script.EnableFileImport(true)
+	}
+	audit.Event(vm.ID.String(), AuditEventName, "Script loaded")
+	return nil
+}
+
+// Compile compiles to byte code loaded copy of vm script
+func (vm *VM) Compile() (err error) {
+	vm.Compiled = new(script.Compiled)
+	vm.Compiled, err = vm.Script.Compile()
+	return
+}
+
+// Run runs byte code
+func (vm *VM) Run() (err error) {
+	audit.Event(vm.ID.String(), AuditEventName, "Script executed")
+	return vm.Compiled.Run()
+}
+
+// RunCtx runs compiled byte code with context.Context support.
+func (vm *VM) RunCtx() (err error) {
+	if vm.ctx == nil {
+		vm.ctx = context.Background()
+	}
+
+	ct, cancel := context.WithTimeout(vm.ctx, GCTScriptConfig.ScriptTimeout)
+	defer cancel()
+
+	audit.Event(vm.ID.String(), AuditEventName, "Script executed")
```

Comment:
Just some ideas for how we can flesh out better auditing, I'd like to see things like:

1) VM Name
2) VM ID
3) Script path
4) sha256sum of file (scripts don't support versioning yet, so we don't know if the contents have been changed between differing script executions)
5) Start time
6) Execution result, success
7) Event (executed)

Reply:
All awesome information to add will start adding increased audit logging 

What are your thoughts on @gloriousCode's comment before about a separate table also for tracking script execution? or would you rather keep it all on the audit table

Reply:
I think a separate table would be good specifically for script events